### DISCUSSION TITLE
Localization support

### DISCUSSION THREAD

**Discussion by guihkx** [Morganamilo/paru#424] [Ideas]
Hello, are there any plans to support other languages in paru? Thanks!

**Comment by crapStone** [comment]
@Morganamilo what do you think about that: https://lib.rs/crates/fluent-bundle

**Comment by Morganamilo** [comment]
I have no idea, I've not looked at localisation libs and when I do I won't be able to get a proper opinion until I play with them a bit.

**Comment by crapStone** [comment]
Should I implement it for some files and you can decide if you want it?

**Comment by Morganamilo** [comment]
Can do. One of my requirements is auto generating templates like .po files.

**Comment by crapStone** [comment]
[Fluent](https://www.projectfluent.org/) is a very new project (the first specification was released in 2019) that aims to provide the best translation possible for each language. But because it is such a young project the ecosystem and especially the Rust implementation isn't very mature yet (the Rust implementation is very low-level and doesn't even support the full specification).

Imho this project is a great new approach to make translations and we could benefit greatly by using it. But because it is still so new there will be probably more than one rewrite of the translation system necessary if we want to use the newest features.

If you don't want this i will use the tried and proven [GNU gettext](https://lib.rs/crates/tr).

**Comment by alerque** [comment]
Even though the ecosystem and library ergonomics are still playing catch up a little bit, Fluent brings *so* much to the table not offered by previous generations of translation systems that I think it's by far the best way to go for a project like this.

I've built several CLI's with Fluent based l10n, if/when work gets started on this I'd be willing to help a little.

**Comment by Morganamilo** [comment]
Do you have an example of such programs? The examples are a little verbose and boilerplate heavy.

**Reply by alerque** [reply]
Sure. Not all of them are public projects, but a couple you can see are `casile` and `fontship`. The implementation in both is roughly the same, both somewhat incomplete because localizing them was just a way to play with Rust. I actually *wouldn't* look at the implementation because they were built with Fluent libraries that are now two major releases old and the ergonomics have improved **a lot**.

Since both tools are quite specialized and it would be hard to see them in action without related projects, here is some sample output from the same build of `casile` on the same demo project files but changing the host system language:

![image](https://user-images.githubusercontent.com/173595/123505699-460b7c00-d669-11eb-80b6-fd23bfb495e8.png)

**Reply by alerque** [reply]
I do suggest if you want better ergonomics for this project that you wait until Clap v3 is released for Rust. Some of the things holding the beta back are related to issues localizing CLIs.

**Reply by Morganamilo** [reply]
I'm interested in the localization code, not the output. Looking at https://github.com/alerque/casile/blob/master/src/i18n.rs there's a lot of boilerplate which makes me want to not use fluent.

**Reply by alerque** [reply]
The boiler plate is mostly because I was flexing and doing weird things like being able to run from EITHER raw `*.ftl` resources *and* compiled into the binary with a fallback system for injecting custom translations at runtime. The other source of boilerplate is locale handling which I hand rolled. The later releases of the Fluent Rust bindings are *much* more ergonomic, and there is a crate for doing most of the boilerplate I invented there.

The Rust side is pretty nice to work with these days. And as a multi-lingual translator Fluent is **so much** better than other localization systems it's not even a contest. The only part that would give me pause is the ecosystem of *translator facing* tools for contributing translations. That's somewhat less developed than other system, although most people that will jump into something like `paru` are going to be able to edit a text file and submit a PR here.

**Reply by Morganamilo** [reply]
Can you point me towards a small example (or implement one yourself against stats.rs)? I'm looking at fluent right now and there''s so many components and I can't seem to find a simple way to just translate strings against a dir of .ftl files.

**Reply by Morganamilo** [reply]
Looking more https://crates.io/crates/i18n-embed-fl seems simple and what I want.

**Reply by Morganamilo** [reply]
Actually it looks that doesn't handle LANG so that's annoying.

**Reply by alerque** [reply]
Unrelated to Fluent, you are going to run up against a dead end using Rust & Clap for the CLI. Some important bits like `--help` output are nearly impossible to localize properly. Some bits are easy, but the 2.x series of Clap releases has a cart-before-the-horse problem that makes localizing all things properly *really hard* no matter what i18n system you use.

As much as it pains me to say it my recommendation is to hold off until Clap resolves this upstream. Otherwise you are destined for a mind-bender to get this working right.

**Reply by Morganamilo** [reply]
Paru does not use clap.

**Reply by alerque** [reply]
My bad, I thought I'd just seen it go by while compiling recently.

**Comment by Silejonu** [comment]
I'd be happy to provide a French localisation once there is a tool in place to contribute translations.

**Reply by Morganamilo** [reply]
Once localization is in place I'll definitely be looking for translators. I'll put a call out here and on reddit when the time comes.

**Comment by juxuanu** [comment]
I'm subscribed to this thread so that I can hop in and translate into Catalan and potentially Spanish once you add support for it.

**Reply by BachoSeven** [reply]
Same here for Italian.

**Comment by Morganamilo** [comment]
I'm probably just going to go for https://docs.rs/tr/. Seems a lot simpler.

**Comment by Morganamilo** [comment]
https://github.com/Morganamilo/paru/pull/428

**Comment by Morganamilo** [comment]
If any one who wants to translate using the `locale` branch just to make sure the system works and doesn't have any kinks feel free. Once I have one translation in place I'll merge it all into master.

**Comment by Morganamilo** [comment]
Everything is now in place. paru-git supports localization and the first translation is now in place thanks to @ouuan.

See #433 or #paru on irc for discussion about translating.

**Comment by Morganamilo** [comment]
Just in case any one is subscribed here but not in the discussion: https://github.com/Morganamilo/paru/discussions/433#discussioncomment-4211504
